In [1]:
import networkx as nx 

G = nx.DiGraph()
G.add_edges_from([('A', 'B'), ('C', 'A'), ('A', 'E'), ('G', 'A'), ('A', 'N'), ('B', 'C'), ('D', 'B'), 
                  ('B', 'E'), ('C', 'D'), ('E', 'C'), ('D', 'E'), ('E', 'D'), ('F', 'G'), ('I', 'F'), 
                  ('J', 'F'), ('H', 'G'), ('I', 'G'), ('G', 'J'), ('I', 'H'), ('H', 'I'), ('I', 'J'), 
                  ('J', 'O'), ('O', 'J'), ('K', 'M'), ('K', 'L'), ('O', 'K'), ('O', 'L'), ('N', 'L'), 
                  ('L', 'M'), ('N', 'O')])

In [3]:
nx.is_strongly_connected(G)

False

In [5]:
nx.connectivity.minimum_edge_cut(G, s="H",t="O")

{('J', 'O'), ('N', 'O')}

In [6]:
karate = nx.karate_club_graph()

In [7]:
karate = nx.convert_node_labels_to_integers(karate, first_label=1)

## Measures of centrality
### Degree centrality =  $\frac{degree}{|N|-1}$
This assumes that important nodes are the ones with the most connections to other nodes

In [10]:
degCentrality = nx.degree_centrality(karate)
degCentrality

{1: 0.48484848484848486,
 2: 0.2727272727272727,
 3: 0.30303030303030304,
 4: 0.18181818181818182,
 5: 0.09090909090909091,
 6: 0.12121212121212122,
 7: 0.12121212121212122,
 8: 0.12121212121212122,
 9: 0.15151515151515152,
 10: 0.06060606060606061,
 11: 0.09090909090909091,
 12: 0.030303030303030304,
 13: 0.06060606060606061,
 14: 0.15151515151515152,
 15: 0.06060606060606061,
 16: 0.06060606060606061,
 17: 0.06060606060606061,
 18: 0.06060606060606061,
 19: 0.06060606060606061,
 20: 0.09090909090909091,
 21: 0.06060606060606061,
 22: 0.06060606060606061,
 23: 0.06060606060606061,
 24: 0.15151515151515152,
 25: 0.09090909090909091,
 26: 0.09090909090909091,
 27: 0.06060606060606061,
 28: 0.12121212121212122,
 29: 0.09090909090909091,
 30: 0.12121212121212122,
 31: 0.12121212121212122,
 32: 0.18181818181818182,
 33: 0.36363636363636365,
 34: 0.5151515151515151}

### Additionaly we could find indegree centrality and out-degree centrality for directed graphs
#### In-degree centrality = $\frac{degree_{in}}{|N| - 1}$
#### Out-degree centrality = $\frac{degree_{out}}{|N| - 1}$

### Closeness centrality of node L = $\frac{|R(L)|}{|N|-1} \frac{|R(L)|}{\sum_{v \in R(L)}{dist(L,v)}}$ 
where R(L) denotes the nodes reachable from L

In [12]:

G1 = nx.read_gml('datadump/friendships.gml')

In [13]:
len(G1.nodes), len(G1.edges)

(1133, 5453)

In [14]:
G1

The graph is a simple graph with 1133 users and 5453 connections betweeen users

### Find the degree centrality, closeness centrality, and normalized betweeness centrality (excluding endpoints) of node 100.


In [16]:
nx.degree_centrality(G1)[100]

0.0026501766784452294

In [18]:
nx.betweenness_centrality(G1,normalized=True,endpoints=False)[100]

7.142902633244772e-05

In [19]:
nx.closeness_centrality(G1,u=100)

0.2654784240150094

#### Q: Suppose you are employed by an online shopping website and are tasked with selecting one user in network G1 to send an online shopping voucher to. We expect that the user who receives the voucher will send it to their friends in the network.  You want the voucher to reach as many nodes as possible. The voucher can be forwarded to multiple users at the same time, but the travel distance of the voucher is limited to one step, which means if the voucher travels more than one step in this network, it is no longer valid. Apply your knowledge in network centrality to select the best candidate for the voucher. 

In [26]:
deg_centrality = nx.degree_centrality(G1)
sorted(deg_centrality.items(),reverse=True, key=itemgetter(1))[:10]

[(105, 0.0636042402826855),
 (23, 0.045936395759717315),
 (333, 0.045936395759717315),
 (16, 0.045053003533568906),
 (42, 0.045053003533568906),
 (41, 0.04328621908127209),
 (196, 0.04151943462897526),
 (233, 0.03975265017667844),
 (21, 0.037985865724381625),
 (76, 0.037985865724381625)]

User **105** is the most ideal to send it to, since they have the highest degree centrality indicating that at 1 hop away they contain the max number of neighbouring nodes

#### Q: Now the limit of the voucher’s travel distance has been removed. Because the network is connected, regardless of who you pick, every node in the network will eventually receive the voucher. However, we now want to ensure that the voucher reaches the nodes in the lowest average number of hops.
How would you change your selection strategy? Write a function to tell us who is the best candidate in the network under this condition.

In [21]:
close = nx.closeness_centrality(G1)

In [24]:
from operator import itemgetter
sorted(close.items(),reverse=True, key=itemgetter(1))[:10]

[(23, 0.3847722637661455),
 (333, 0.3828204261075414),
 (105, 0.38063214525891054),
 (42, 0.3775850567044696),
 (41, 0.3749585955614442),
 (76, 0.37433862433862436),
 (233, 0.3733509234828496),
 (52, 0.37322782723376197),
 (135, 0.36993464052287583),
 (378, 0.36860957342885053)]

User **23** seems to be a good candidate to send the voucher to, as they have the highest closeness centrality indicating that they would be able to reach other nodes with the shortest paths overall

#### `G2` is a directed network of political blogs, where nodes correspond to a blog and edges correspond to links between blogs.

In [29]:
G2 = nx.read_gml('datadump/blogs.gml')

In [30]:
len(G2.nodes), len(G2.edges)

(1490, 19025)

In [31]:
G2

#### Find top 5 nodes with highest scaled pagerank

In [33]:
pg_rank = nx.pagerank(G2,alpha=0.85)

In [37]:
dict(sorted(pg_rank.items(),reverse=True, key=itemgetter(1))[:5])

{'dailykos.com': 0.017901443885198386,
 'atrios.blogspot.com': 0.015178631721614684,
 'instapundit.com': 0.012627090660729751,
 'blogsforbush.com': 0.012508582138399098,
 'talkingpointsmemo.com': 0.012393033204751033}

In [38]:
pg_rank["realclearpolitics.com"]

0.004636694781649094

#### Find top 5 hubs and authorities

In [39]:
hubs, authorities = nx.hits(G2,normalized=True)

In [40]:
def get_top_N(dictionary, N):
    return dict(sorted(dictionary.items(),reverse=True, key=itemgetter(1))[:N])

In [41]:
get_top_N(hubs, 5)

{'politicalstrategy.org': 0.006860032844631305,
 'madkane.com/notable.html': 0.006198130021198813,
 'liberaloasis.com': 0.0061346896013232885,
 'stagefour.typepad.com/commonprejudice': 0.0059907290973003745,
 'bodyandsoul.typepad.com': 0.005939626690753934}

In [42]:
get_top_N(authorities, 5)

{'dailykos.com': 0.015042267072353255,
 'talkingpointsmemo.com': 0.014450907816427404,
 'atrios.blogspot.com': 0.014083800022788357,
 'washingtonmonthly.com': 0.011953445820310986,
 'talkleft.com': 0.009705131061986168}

In [43]:
hubs["realclearpolitics.com"], authorities["realclearpolitics.com"]

(0.0003243556140916672, 0.003918957645699851)